lib/datadog/core/tag_normalizer.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+module Datadog
+  module Core
+    module TagNormalizer
+      # Normalization logic used for tag keys and values that the Trace Agent has for traces
+      # Useful for ensuring that tag keys and values are normalized consistently
+      # An use case for now is Process Tags which need to be sent across various intakes (profiling, tracing, etc.) consistently
+
+      module_function
+
+      INVALID_TAG_CHARACTERS = %r{[^\p{L}0-9_\-:./]}
+      LEADING_INVALID_CHARS_NO_DIGITS = %r{\A[^\p{L}:]++}
+      LEADING_INVALID_CHARS_WITH_DIGITS = %r{\A[^\p{L}0-9:./\-]++}
+      MAX_BYTE_SIZE = 200 # Represents the max tag length
+      TRAILING_UNDERSCORES = %r{_++\z}
+      VALID_ASCII_TAG = %r{\A[a-z:][a-z0-9:./-]*\z}
+
+      # Based on https://github.com/DataDog/datadog-agent/blob/45799c842bbd216bcda208737f9f11cade6fdd95/pkg/trace/traceutil/normalize.go#L131
+      # Specifically:
+      # - Must be valid UTF-8
+      # - Invalid characters are replaced with an underscore
+      # - Leading non-letter characters are removed but colons are kept
+      # - Trailing non-letter characters are removed
+      # - Trailing underscores are removed
+      # - Consecutive underscores are merged into a single underscore
+      # - Maximum length is 200 characters
+      # If it's a tag value, allow it to start with a digit
+      # @param original_value [String] The original string
+      # @param remove_digit_start_char [Boolean] - whether to remove the leading digit (currently only used for tag values)
+      # @return [String] The normalized string
+      def self.normalize(original_value, remove_digit_start_char: false)
+        transformed_value = original_value.to_s.encode('UTF-8', invalid: :replace, undef: :replace)
```

Comment:
I think there is a function for this in `Core::Utils#utf8_encode`

Reply:
I didn't realize `utf8_encode` existed and had to add a small change to be able to use it: https://github.com/DataDog/dd-trace-rb/pull/5033/commits/e83bc4a9e498debdbc2c6d1dba3d1a5e7863d63a.